File: command/java/CommandPatternDemo.java
import java.util.List;
import java.util.Vector;

interface OrderCommand {

	void execute();
}

final class Stock {

	private String name;
	private int quantity;

	public Stock(String name, int quantity) {
		this.name = name;
		this.quantity = quantity;
	}

	public String getName() {
		return this.name;
	}

	public int getQuantity() {
		return this.quantity;
	}

	public void buy() {
		System.out.printf("Stock [name: %s, quantity: %d] bought\n", this.name, this.quantity);
	}

	public void sell() {
		System.out.printf("Stock [name: %s, quantity: %d] sold\n", this.name, this.quantity);
	}
}

final class BuyStock implements OrderCommand {

	private Stock stock;

	public BuyStock(Stock stock) {
		this.stock = stock;
	}

	@Override
	public void execute() {
		stock.buy();
	}
}

final class SellStock implements OrderCommand {

	private Stock stock;

	public SellStock(Stock stock) {
		this.stock = stock;
	}

	@Override
	public void execute() {
		stock.sell();
	}
}

final class Broker {

	private final List<OrderCommand> orders;

	public Broker() {
		this.orders = new Vector();
	}

	public void takeOrder(OrderCommand order) {
		this.orders.add(order);
	}

	public void placeOrders() {
		this.orders.stream().forEach(OrderCommand::execute);
		this.orders.clear();
	}
}

final class CommandPatternDemo {

	public static void main(String[] args) {
		Stock stock = new Stock("Shoe", 596);
		OrderCommand buyCommand = new BuyStock(stock);
		OrderCommand sellCommand = new SellStock(stock);
		Broker broker = new Broker();

		broker.takeOrder(buyCommand);
		broker.takeOrder(sellCommand);

		broker.placeOrders();
	}
}

File: command/kotlin/command_pattern_demo.kt
interface OrderCommand {

	fun execute(): Unit
}

class Stock(
	private val name: String,
	private val quantity: Int
) {

	fun buy(): Unit {
		println("Stock [name: ${this.name}, quantity: ${this.quantity}] bought")
	}

	fun sell(): Unit {
		println("Stock [name: ${this.name}, quantity: ${this.quantity}] sold")
	}
}

class BuyStock(private val stock: Stock) : OrderCommand {

	override fun execute(): Unit {
		stock.buy();
	}
}

class SellStock(private val stock: Stock) : OrderCommand {

	override fun execute(): Unit {
		stock.sell();
	}
}

class Broker(private val orders: MutableList<OrderCommand>) {

	constructor() : this(mutableListOf())

	fun takeOrder(order: OrderCommand): Unit {
		this.orders.add(order);
	}

	fun placeOrders(): Unit {
		this.orders.forEach(OrderCommand::execute);
		this.orders.clear();
	}
}

fun main(): Unit {
	val stock: Stock = Stock("Shoe", 596);
	val buyCommand: OrderCommand = BuyStock(stock);
	val sellCommand: OrderCommand = SellStock(stock);
	val broker: Broker = Broker();

	broker.takeOrder(buyCommand);
	broker.takeOrder(sellCommand);

	broker.placeOrders();
}